services/py-api/src/database/repository/participants_repository.py

```diff
@@ -0,0 +1,54 @@
+from typing import Optional, Type
+
+from bson import ObjectId
+from motor.motor_asyncio import AsyncIOMotorClientSession
+from pydantic import BaseModel
+from pymongo.errors import DuplicateKeyError
+from result import Result, Ok, Err
+from structlog.stdlib import get_logger
+
+from src.database.db_manager import DatabaseManager
+from src.database.model.participant_model import Participant
+from src.database.repository.base_repository import CRUDRepository
+from src.server.exception import DuplicateEmail
+from src.server.schemas.request_schemas.schemas import ParticipantRequestBody
+
+LOG = get_logger()
+
+
+class ParticipantsRepository(CRUDRepository):
+    # TODO: Implement the rest of the methods
+    def __init__(self, db_manager: DatabaseManager, collection: str) -> None:
+        self._client = db_manager.client
+        self._collection = self._client[db_manager.DB_NAME][collection]
+
+    async def fetch_by_id(self, obj_id: str) -> Result:
+        raise NotImplementedError()
+
+    async def fetch_all(self) -> Result:
+        raise NotImplementedError()
+
+    async def update(
+        self, obj_id: str, input_data: BaseModel, session: Optional[AsyncIOMotorClientSession] = None
+    ) -> Result:
+        raise NotImplementedError()
+
+    async def delete(self, obj_id: str, session: Optional[AsyncIOMotorClientSession] = None) -> Result:
+        raise NotImplementedError()
+
+    async def create(
+        self, input_data: ParticipantRequestBody, session: Optional[AsyncIOMotorClientSession] = None
```

Comment:
What's the whole deal with async in the first place?

Reply:
As we have async handling of requests it makes sense to use an async library, in this case `motor`. Otherwise the sync operations of the normal `pymongo` library would block the event loop.

Reply:
What async handling do you have?

Reply:
FastAPI as a whole is async. I can't understand your question.

Reply:
Yes. FastAPI is built around async but doesn't force you into using async. Asynchronous code is resource demanding and doesn't make sense to use it if you don't actually need it.

Reply:
Also, async in this context just adds an additional layer of complexity.

Reply:
The only functionality I can think of that might possibly need an async operation would be the functionality around sending emails.

Reply:
What about db queries? These are also IO bound operations. In JS I have seen db libraries which are async.

Reply:
Yes. Async in JavaScript is a native concept. But let's first take a step back and figure out what async really means. Async is like running operations in parallel, right? Why do you need to overwhelm your code with async resolutions and event loop processing if your code doesn't have to run multiple independent of each other operations at once? 

Think of the following case. You want to create 5 team members at once and add all of them to a team. This makes sense to be an async operation. You batch these 5 queries and execute them in "parallel" context, when all of these are resolved, take the ids and insert them into team entity. 

In your case, I'm not saying to totally get rid of async but instead bring it's usage to a minimum if you don't really need it. I can't think of operations in your business use cases where you would need such an async operation where you want to keep running the thread without blocking it.

Reply:
@asynchroza Ok, from my knowledge and [this MDN article ](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Introducing) async is primarily used when you have IO-bound operations, which could take some time. In our case, a db query could take some time, but we don't want to wait for the query to finish (block the thread) before accepting the next client request. 

Enough about theory, the best way to see which approach is better is to load test the API, so that's what we did with @Isak-Bego. 
The tests were performed on Isak's MacBook Pro (M3 pro, not sure about the concrete one and 18 GB of RAM)

We used a single `uvicorn` worker and we left the config of the connection pool with a minimum of 25 idle connections. Our MongoDB instance is deployed in AWS Frankfurt (eu-central-1).

For the load test, we used `locust` with a max concurrency (number of users) of 200 and a ramp-up rate of 10. Each test was performed 3 times.  Each test was performed on a clean db (0 participants and 0 teams).

The 2 screenshots below are from each group of tests, and the numbers I have written are the average from the three tests performed for each test case.

**The first test case was performed with the current state of the code (full async and `motor` lib):**


<img width="1267" alt="image" src="https://github.com/user-attachments/assets/05963725-1f19-4720-bb24-d349996a774d">
<img width="1161" alt="image" src="https://github.com/user-attachments/assets/07f45eae-7715-4008-a1a4-42e70258dcd0">

Average response time in ms from the three tests: 3593.29
Average number of failed requests from the three tests: 0.33 (We had only one failed request for all three tests):
Cause of failure:

<img width="1156" alt="image" src="https://github.com/user-attachments/assets/1e9f95bf-e573-40a3-befb-17f23187da29">

-----------------------------------------------------------------------------------------------------------------

**The second test case was performed with no async anywhere using the `pymongo` library.**


<img width="1208" alt="image" src="https://github.com/user-attachments/assets/24e3df0b-18f9-40b0-a272-9a195f0bfd8a">
<img width="1180" alt="image" src="https://github.com/user-attachments/assets/b7acbdc7-1e01-490f-8c56-e2af12c39cc3">

Average response time in ms from the three tests: 4037.56
Average number of failed requests from the three tests: 1.33 (We had a total of  failed 4 requests for all tests):
Cause of failure:

<img width="1160" alt="image" src="https://github.com/user-attachments/assets/7cbaecdc-c7ae-4374-a13d-80194f3df62c">

-----------------------------------------------------------------------------------------------------------------

**BONUS: The third test case was performed with async only on the route using the `pymongo` library. The test was performed only once with a max concurrency of 100 users and a ramp-up of 10**
<img width="853" alt="image" src="https://github.com/user-attachments/assets/bee3d0f2-7d14-4a63-a2d1-d40c57852d38">

Here, we wanted to see if the event loop would really get blocked (it did), as we are executing a synchronous operation in an async context.

<img width="1149" alt="image" src="https://github.com/user-attachments/assets/ea86fa85-9898-4b11-8ca8-5741cff74a87">
<img width="1152" alt="image" src="https://github.com/user-attachments/assets/01e92cd2-ba8f-4382-bc6a-3779e0186a2f">

-----------------------------------------------------------------------------------------------------------------

Summary: 
**We are 444,27 ms faster on average in terms of response time, when we use the async Mongo library 

**`motor`.** And we have a lower number of failed requests, though when taking the average, it's only 1.**

Reply:
Thank you for taking the time to analyze this and compile the report, it's amazing and proves I was mistaken.

Upon revisiting my original comment, I realize I was primarily focused on how async affects a single method and its I/O dependency, rather than considering the broader global context. As we discussed in the chat, the choice here seems to be between prioritizing a highly efficient API or one that’s easier for newcomers to adopt.

Reply:
Thank you Misho, means a lot from you!